Incorporate MT's design (#42)
quicwg/version-negotiation: draft-ietf-quic-version-negotiation.md

when, and only when, they appear in all capitals, as shown here.

## Server Deployments of QUIC {#server-terms}

While this document mainly discusses a single QUIC server, it is common for
deployments of QUIC servers to include a fleet of multiple servers. We
therefore define the following terms:

Supported Versions:

> : This is the set of versions supported by a given individual server.

Reviewer: ```suggestion
: This is the set of versions supported by a given individual server instance.
```

While this document mainly discusses a single QUIC server, it is common for
deployments of QUIC servers to include a fleet of multiple server instances. We
therefore define the following terms:

Accepted Versions:

: This is the set of versions supported by a given server instance. More
specifically, these are the versions that a given server instance will use if a
client sends a first flight using them.

Negotiated Versions:

: This is the set of versions that a given server instance will send in a
Version Negotiation packet if it receives a first flight from an unknown
version. This set will most often be equal to the Accepted Versions set, except
during short transitions while versions are added or removed (see below).

Fully-Deployed Versions:

: This is the set of QUIC versions that is supported and negotiated by every
single QUIC server instance in this deployment. If a deployment only contains a
single server instance, then this set is equal to the Negotiated Versions set,
except during short transitions while versions are added or removed (see below).

If a deployment contains multiple server instances, it is possible for software

> updates to not happen at the exact same time on all server instances. Because

Reviewer 2: ```suggestion
updates to happen a few server instances at a time. Because
```
Removes a split infinitive and language that feels excessively informal ("exact same time").

of this, a client might receive a Version Negotiation packet from a server
instance that has already been updated and the client's resulting connection
attempt might reach a different server instance which hasn't been updated yet.

However, even when there is only a single server instance, it is still possible
to receive a stale Version Negotiation packet if the server performs its
software update while the Version Negotiation packet is in flight.

This could cause the version downgrade prevention mechanism described in
{{downgrade}} to falsely detect a downgrade attack. To avoid that, server
operators SHOULD perform a three-step process when they wish to add or remove

> support for a version:
> 
> * When adding support for a new version, the first step is to progressively add
>   support for the new version to all server instances. This step updates the
>   Accepted Versions but not the Negotiated Versions nor the Fully-Deployed
>   Versions. Once all server instances have been updated, operators wait for at
>   least one minute to allow any in-flight Version Negotiation packets to
>   arrive. Then, the second step is to progressively add the new version to
>   Negotiated Versions on all server instances. Once complete, operators wait
>   for at least another minute. Finally, the third step is to progressively add
>   the new version to Fully-Deployed Versions on all server instances.

Reviewer 2: ```suggestion
support for a version.

When adding support for a new version:

- The first step is to progressively add
  support for the new version to all server instances. This step updates the
  Accepted Versions but not the Negotiated Versions nor the Fully-Deployed
  Versions. Once all server instances have been updated, operators wait for at
  least one minute to allow any in-flight Version Negotiation packets to
  arrive. 
- Next, progressively add the new version to
  Negotiated Versions on all server instances. Once complete, operators wait
  for at least another minute.
- Finally, progressively add
  the new version to Fully-Deployed Versions on all server instances.
```
Modulo rewrapping, I think moving the bullets down a level is clearer.

Author: Fixed

> * When removing support for a version, the first step is to progressively
>   remove the version from Fully-Deployed Versions on all server instances. Once
>   it has been removed on all server instances, operators wait for at least one
>   minute to allow any in-flight Version Negotiation packets to arrive. Then,
>   the second step is to progressively remove the version from Negotiated
>   Versions on all server instances. Once complete, operators wait for at least
>   another minute. Finally, the third step is to progressively remove support for
>   the version from all server instances. That step updates the Supported
>   Versions.

Reviewer 2: ```suggestion
When removing support for a version:

- The first step is to progressively
  remove the version from Fully-Deployed Versions on all server instances. Once
  it has been removed on all server instances, operators wait for at least one
  minute to allow any in-flight Version Negotiation packets to arrive.
 - Next, progressively remove the version from Negotiated
  Versions on all server instances. Once complete, operators wait for at least
  another minute.
- Finally, progressively remove support for
  the version from all server instances. That step updates the Supported
  Versions.
```
...and the same down here.

Author: Fixed

when, and only when, they appear in all capitals, as shown here.

# Server Deployments of QUIC {#server-fleet}

While this document mainly discusses a single QUIC server, it is common for
deployments of QUIC servers to include a fleet of multiple server instances. We
therefore define the following terms:

Supported Versions:

: This is the set of versions supported by a given individual server instance.

Fully-Deployed Versions:

: This is the set of QUIC versions that is supported by every single QUIC
server instance in this deployment.

Partially-Deployed Versions:

: This is the set of versions supported by a given individual server instance,
but which aren't supported by every single QUIC server instance in this
deployment.

If a deployment only contains a single server instance, then the Supported

> Versions is equal to the Fully-Deployed Versions, and the Partially-Deployed

Reviewer 3: nit: "Supported Versions" is plural, but "is" is singular. You might consider using "are" instead of "is" or alternatively dodging by using "Supported Versions set" (which is singular).

Versions is the empty set. Note however that during software upgrades these

> fields are upgraded at different times, as described below.

Reviewer 3: nit: "fields" or "sets"

to receive a stale Version Negotiation packet if the server performs its
software update while the Version Negotiation packet is in flight.

In either of these events, when the client reacts to a Version Negotiation
packet, it might create a new QUIC connection which could reach a server
instance which has a different set of Supported Versions than what was received
in the Version Negotiation packet.

This could cause the version downgrade prevention mechanism described in
{{downgrade}} to falsely detect a downgrade attack. To avoid that, server
operators MUST perform a two-step process when they wish to add or remove
support for a version:

* When adding support for a new version, the first step is to progressively add
  support for the new version to all server instances. This step updates the
  Supported Versions but not the Fully-Deployed Versions. Once all server
  instances have been upgraded, operators wait for at least one minute to allow
  any in-flight Version Negotiation packets to arrive. Then, the second step is
  to progressively add the new version to Fully-Deployed Versions on all server

>   instances.

Reviewer: Do you add the code AND start sending VN with the new version at the same time?  Because sending VN in that state will result in connection failures.  As I said, this probably needs a three step dance.

Author: No, the key property here is that VN packets carry the Fully-Deployed Versions, so that happens after the 1min wait. Or am I missing something?

Reviewer: So the problem is that if you change VN and Other Versions at the same time. You will have a mismatch between VN and Other Versions in both directions.  Some clients will see VN without the new version and OV with.  Other clients will see VN with the new version and OV without.

It's that former of those two that creates a problem; if the client sees VN without version X and then sees version X in OV, then it will fail the connection.

Author: /facepalm. Yup, you're right. Three steps it is.

when, and only when, they appear in all capitals, as shown here.

# Server Deployments of QUIC {#server-fleet}

While this document mainly discusses a single QUIC server, it is common for
deployments of QUIC servers to include a fleet of multiple server instances. We
therefore define the following terms:

> Accepted Versions:

Reviewer 4: Nit: I would call this "acceptable"

: This is the set of versions supported by a given server instance. More
specifically, these are the versions that a given server instance will use if a
client sends a first flight using them.

> Negotiated Versions:

Reviewer 4: And I would call this "offered versions"

has already been updated and the client's resulting connection attempt might
reach a different server instance which hasn't been updated yet.

However, even when there is only a single server instance, it is still possible
to receive a stale Version Negotiation packet if the server performs its
software update while the Version Negotiation packet is in flight.

This could cause the version downgrade prevention mechanism described in
{{downgrade}} to falsely detect a downgrade attack. To avoid that, server
operators SHOULD perform a three-step process when they wish to add or remove
support for a version:

When adding support for a new version:

* The first step is to progressively add support for the new version to all
  server instances. This step updates the Accepted Versions but not the
  Negotiated Versions nor the Fully-Deployed Versions. Once all server
  instances have been updated, operators wait for at least one minute to allow
  any in-flight Version Negotiation packets to arrive.

* Then, the second step is to progressively add the new version to Negotiated
  Versions on all server instances. Once complete, operators wait for at least

>   another minute.

Reviewer 4: This should be phrased in terms of MSL.

* Finally, the third step is to progressively add the new version to
  Fully-Deployed Versions on all server instances.

When removing support for a version:

* The first step is to progressively remove the version from Fully-Deployed
  Versions on all server instances. Once it has been removed on all server
  instances, operators wait for at least one minute to allow any in-flight

>   Version Negotiation packets to arrive.

Reviewer 4: What does this mean operationally. Fully-Deployed is defined as being across all the servers.

Author: The moment you remove it from the list it means that server no longer sends it in the TP.

Reviewer 4: How is that different from removal from Negotiated versions

Author: Negotiated Versions is sent in VN packets, these are sent in the TP. They're different.

server.

# Server Deployments of QUIC {#server-fleet}

While this document mainly discusses a single QUIC server, it is common for
deployments of QUIC servers to include a fleet of multiple server instances. We
therefore define the following terms:

Supported Versions:

: This is the set of versions supported by a given individual server instance.

Fully-Deployed Versions:

: This is the set of QUIC versions that is supported by every single QUIC
server instance in this deployment.

Partially-Deployed Versions:

: This is the set of versions supported by a given individual server instance,
but which aren't supported by every single QUIC server instance in this
deployment.

If a deployment only contains a single server instance, then the Supported
Versions is equal to the Fully-Deployed Versions, and the Partially-Deployed
Versions is the empty set.

Conversely, if a deployment contains multiple server instances, it is possible
for software updates to not happen at the exact same time on all server

> instances. In this scenario, some instances will have a different Supported

Reviewer: I think that you can just say that an instantaneous upgrade is impossible.  A server deployment that includes multiple server instances cannot upgrade all instances instantaneously.  Even if that were possible, there could still be outdated Version Negotiation packets in flight that might result in negotiation failure if downgrade protection were engaged atomically.

This also applies to a single instance deployment.  Though you might avoid the problem by ensuring that your upgrade involves enough downtime to avoid the in-flight VN problem, you might as well just apply the same logic as a multi-server deployment.

Based on that I would remove the "single server instance" paragraph and replace it with a description of the in-flight VN problem.

Versions from other instances in the same fleet. When that happens, a client
might receive a Version Negotiation packet from a server instance that doesn't
support some Partially-Supported Versions; in that case, those
Partially-Supported Versions won't be listed in the Version Negotiation packet.
When the client reacts to that Version Negotiation packet, it might create a
new QUIC connection which could be routed to a different server instance, which
has a different set of Supported Versions. The version downgrade prevention
mechanism described in {{downgrade}} is robust to this scenario and will not
falsely detect a downgrade attack.

> If a server operator is progressively deploying a new QUIC version throughout

Reviewer: Based on the above, I would remove the condition here: just describe how a new version is deployed.

its fleet, it MUST perform a two-step process where it first progressively adds
support for the new version. This adds the new version to Supported Versions
but not to Fully-Deployed Versions. Once all server instances have been
upgraded, the second step is to progressively add the version to Fully-Deployed

> Versions on all server instances. This opens connections to version downgrades

Reviewer: This needs to be three steps:

1. Add code for version X
2. Advertise version X in Version Negotiation
3. Add version X to Other Versions

The reason I say this is the in-flight VN.  Let's say that you support version X and you just add it both VN and OV.  What about clients that 2ms ago got a VN without X in it?

The transition between step 2 and 3 can be short.  You can run that on a fairly short timer for a single server deployment, but you need to ensure that all VNs are out of the system before you move to step 3 or connections will fail.

The logic here is:

* don't update the advisory signals (VN) until the code is in place
* don't update the downgrade protection (OV) until all the advisory signals are in place

Rollback is similar: first remove downgrade protection, then remove advisory signals, then remove actual support.

One thing I like about this design is that we because we don't rely on VN being correct, the design can be made robust against VN being wrong for a short period.

See also https://quicwg.org/ops-drafts/draft-ietf-quic-applicability.html#name-enabling-new-versions

## Incompatible Version Negotiation {#incompat-vn}

The server starts incompatible version negotiation by sending a Version

> Negotiation packet. This packet SHALL list the server's set of Negotiated
> Versions (see {{server-fleet}}) in the packet's Supported Version field. The

Reviewer 2: Nit:  Supported Version doesn't contain a list -- it contains a single version and repeats as many times as necessary to convey the list.

```suggestion
Negotiation packet. This packet SHALL include each entry from the server's
set of Negotiated Versions (see {{server-fleet}}) in a Supported Version
field. The
```

Author: Fixed.

During the handshake, endpoints will exchange handshake version information,
which is a blob of data that is defined below. In QUIC version 1, the handshake
version information is transmitted using a new transport parameter,
`version_negotiation`. The contents of handshake version information are shown
below (using the notation from the "Notational Conventions" section of
{{QUIC}}):

~~~
Handshake Version Information {

>   Currently Attempted Version (32),

Reviewer: I think that this might be better named "Chosen Version".  It's not necessarily the first choice, but the choice that is made after seeing what the other side supports (the client chooses by guessing if it doesn't get VN, and chooses from VN if it does; the server chooses from compatible versions).

Author: Agreed, done.

  Other Versions (32) ...,
}
~~~
{: #fig-hvi-format title="Handshake Version Information Format"}

The content of each field is described below:

Currently Attempted Version:

: This field SHALL be equal to the value of the `Version` field in the long

> header that carries this data.

Reviewer: I'm not sure that this is the definition we want.  I *think* that you are right that this matches the version in the long header, but that is an accident of the design.  The key is that this is the version that the endpoint chose.

It might be possible with retransmissions of Initial packets to include this information even after the server - and server - has switched to a compatible version.  While it might be obvious that the server got the info, the retransmission machinery on the client might not have received an ACK and so it could send this again.

Author: Agreed, I tweaked the text.

The contents of the `Other Versions` field depends on whether it is sent by the
client or by the server.

Client-Sent Other Versions:

: When sent by a client, the `Other Versions` field lists all the versions that
this first flight is compatible with, ordered by descending preference. Note
that the version in the `Currently Attempted Version` field MUST be included in
this list to allow the client to communicate the currently attempted version's
preference. Note that this preference is only advisory, servers MAY choose to

> use their own preference instead.

Reviewer: I'm still not on board with preference ordering.  But let's track that separately.

Author: I filed #43 to track this.

Server-Sent Other Versions:

: When sent by a server, the `Other Versions` field lists all the
Fully-Supported Versions of this server deployment, see {{server-terms}}. This
field conveys Fully-Supported Versions instead of Supported Versions to allow
scenarios where the client's first flight (before receiving VN) reached a

> different server than the client's subsequent flight (after receiving VN).

Reviewer: You might need to expand on this, as it is a key concept.  

This might be worth an entire section that says something like:

The client might receive a Version Negotiation packet from a server instance that doesn't support some Partially-Supported Versions; if it does so, those Partially-Supported Versions won't be listed in the Version Negotiation packet.  As part of validating the server's Other Versions, the client verifies that it would not have chosen any of the versions the server listed.  If the client supports any Partially-Supported Version and would prefer that version, it will detect a downgrade attack.  To avoid this, the server does not advertise any version that is Partially-Supported.

(And then we need text about how to turn on new versions properly.  (1) Deploy code and accept connections; (2) turn on VN for that version; (3) turn on Other Versions for that version and enjoy downgrade protection.)

Author: Agreed, fleshed out the "Server Deployments of QUIC" section

Clients and servers MAY both include versions following the pattern
`0x?a?a?a?a` in their `Other Versions` list. Those versions are reserved to
exercise version negotiation (see the Versions section of {{QUIC}}), and will

> be ignored when parsing these fields.

Reviewer: These aren't ignored, it's just that they are never selected.

Author: Fixed

The new first flight will allow the endpoints to establish a connection using
the negotiated version. The handshake of the negotiated version will exchange

> Version Information (see {{vers-info}}) required to ensure that VN was genuine,

Reviewer 2: Capitalizing this suggests it's referring to the structure, but it's the information within that's relevant.
```suggestion
information (see {{vers-info}}) used to ensure that VN was genuine,
```

Author: Fixed

i.e. that no attacker injected packets in order to influence the VN process,
see {{downgrade}}.

## Compatible Version Negotiation {#compat-vn}

When the server can parse the client's first flight using the original version,

> it can extract the client's Version Information (see {{vers-info}}). This

Reviewer 2: ```suggestion
it can extract the client's Version Information structure (see {{vers-info}}). This
```

Author: Fixed

~~~
Handshake Version Information {
  Chosen Version (32),
  Other Versions (32) ...,
}
~~~
{: #fig-hvi-format title="Handshake Version Information Format"}

The content of each field is described below:

Chosen Version:

: The version that the sender has chosen to use for this connection. This field
MUST be equal to the value of the Version field in the long header that carries

> this data.

Reviewer: As I said before, I think that you can - at best - say that the field is generally equal to the Version field in the packet that carries the data.  You can't use "MUST".

Reviewer: So the server doesn't need to validate what the client provides (it is picking from the list, and **use** >> **validation**).

The client verifies that (a) the version it sent prior to receiving a Version Negotiation packet is not listed, and (b) it would not have selected a different version if it knew what versions the server truly supports.  Note here that the client's choice is based on what the server signals in other versions, the chosen version (if we drop ordering, as we should).

I realize now that the server can't drop compatible versions from this list.  There's a long and complicated argument for why that doesn't work if you are interested, but I need to eat a well-overdue lunch and it might be better over a video link because it's tricky.

Author: > So the server doesn't need to validate what the client provides (it is picking from the list, and **use** >> **validation**).

I've tweaked the text to clarify that this is the version used, not the one from long headers. That allowed me to drop this validation.

> The client verifies that (a) the version it sent prior to receiving a Version Negotiation packet is not listed, and (b) it would not have selected a different version if it knew what versions the server truly supports.

Oh good catch I was missing (b).

> Note here that the client's choice is based on what the server signals in other versions, the chosen version (if we drop ordering, as we should).

Note that in the current proposal we have ordering of the client's versions but not of the server's versions.

> I realize now that the server can't drop compatible versions from this list. There's a long and complicated argument for why that doesn't work if you are interested, but I need to eat a well-overdue lunch and it might be better over a video link because it's tricky.

In the current proposal endpoints can disagree on which versions are compatible so we need them in the list in case the client doesn't think they are compatible.

Reviewer: That is not the reason I had, but as the server generates its list after the client, it would know what the client thinks is compatible when generating its list.

Author: Oh that's a good point. I'd love to hear the full story over VC one of these days then, but no rush.

But overall, do you think this PR is ready to be merged? (If you do I'll email the list to get WG input)

server.

# Version Information {#vers-info}

During the handshake, endpoints will exchange Version Information, which is a
blob of data that is defined below. In QUIC version 1, the Version Information
is transmitted using a new transport parameter, `version_information`. The

> contents of Version Information are shown below (using the notation from the
> "Notational Conventions" section of {{QUIC}}):

Reviewer 2: ```suggestion
contents of Version Information are shown below (using the notation from
{{Section 1.3 of QUIC}}):
```

  Other Versions (32) ...,
}
~~~
{: #fig-vi-format title="Version Information Format"}

The content of each field is described below:

Chosen Version:

: The version that the sender has chosen to use for this connection. In most
cases, this field will be equal to the value of the Version field in the long
header that carries this data.

The contents of the `Other Versions` field depends on whether it is sent by the
client or by the server.

Client-Sent Other Versions:

: When sent by a client, the `Other Versions` field lists all the versions that
this first flight is compatible with, ordered by descending preference. Note
that the version in the `Chosen Version` field MUST be included in this list to
allow the client to communicate the chosen version's preference. Note that this
preference is only advisory, servers MAY choose to use their own preference
instead.

Server-Sent Other Versions:

: When sent by a server, the `Other Versions` field lists all the
Fully-Supported Versions of this server deployment, see {{server-fleet}}.

Clients and servers MAY both include versions following the pattern
`0x?a?a?a?a` in their `Other Versions` list. Those versions are reserved to

> exercise version negotiation (see the Versions section of {{QUIC}}), and will

Reviewer 2: ```suggestion
exercise version negotiation (see {{Section 15 of QUIC}}), and will
```

During the handshake, endpoints will exchange handshake version information,
which is a blob of data that is defined below. In QUIC version 1, the handshake
version information is transmitted using a new transport parameter,
`version_negotiation`. The contents of handshake version information are shown
below (using the notation from the "Notational Conventions" section of

> {{QUIC}}):

Reviewer: You can use `{{Section 15 of QUIC}}` if you like, or even `{{Section 15 (Versions) of QUIC}}`.  Both work (if you have a recent kramdown version).

Author: `{{Section 1.3 of QUIC}}` works but `{{Section 1.3 (Notational Conventions) of QUIC}}` doesn't. I'd rather have the section title when the other document isn't yet published because it could get renumbered.

FWIW it's failing with `Error: IDREF attribute target references an unknown ID "Section_1.3__Notational_Conventions__of_QUIC", at None` and I'm at https://github.com/martinthomson/i-d-template/commit/4639e5ce1002e49ee93a4bb3d9acd7644b3a83de.

Author: Oh I'm realizing you said a recent kramdown, not a recent i-d-template. I'll look into updating that.

~~~

> Handshake Version Information {

Reviewer: How about just "Version Information"?

Author: I like that, done.

> : When sent by a client, the `Other Versions` field lists all the versions that

Reviewer: I just noticed this editorial convention: I think that you should avoid using backticks for field names.  The QUIC spec doesn't use them.  (One nasty consequence of using them is that you get scare quotes in the text rendering.)

connection. If the connection was using QUIC version 1, it MUST be closed with
a transport error of type `VERSION_NEGOTIATION_ERROR`.

If a client has reacted to a Version Negotiation packet, it MUST parse the
server's `Other Versions` field and validate that it does not contain the

> client's original version. If this validation fails, the server MUST close the

Reviewer: ```suggestion
client's original version. If this validation fails, the client MUST close the
```

connection. If the connection was using QUIC version 1, it MUST be closed with

> a transport error of type `VERSION_NEGOTIATION_ERROR`. This mitigates an
> attacker's ability to forge Version Negotiation packets to force a version
> downgrade.

Reviewer: ```suggestion
a transport error of type `VERSION_NEGOTIATION_ERROR`. This prevents 
an attacker from being able to use forged Version Negotiation packets to
force a version downgrade.
```

Negotiation packet, it MUST drop all subsequent Version Negotiation packets on
that connection.

Both endpoints MUST parse their peer's Version Information during the
handshake. If the Version Information was missing or if parsing it failed (for
example, if it is too short or if its length is not divisible by four), then
the endpoint MUST close the connection. If the connection was using QUIC
version 1, it MUST be closed with a transport error of type
`TRANSPORT_PARAMETER_ERROR`.

If a client has reacted to a Version Negotiation packet, it MUST validate that
the server's `Other Versions` field does not contain the client's original
version, and that the client would have selected the same negotiated version if
it had received a Version Negotiation packet whose Supported Versions field had
the same contents as the server's `Other Versions` field. If any of these
validations fail, the client MUST close the connection. If the connection was
using QUIC version 1, it MUST be closed with a transport error of type
`VERSION_NEGOTIATION_ERROR`. This prevents an attacker from being able to use
forged Version Negotiation packets to force a version downgrade.

> 

Reviewer 5: There is a tradeoff here between "punish the attacker" and "punish the client". Suppose that when all is said and done, the server proposes version X. Suppose also that version X is equally or better preferred by the client over the original version Y and the version Z proposed after version negotiation. For all practical matter, the attack has been thwarted, and the negotiation landed in a good place. Do you really want the client to discard the connection? Isn't that providing VN packets with the power of DOS?

Reviewer 5: On the other hand, I do like putting the onus of verification on the client. That's a significant improvement over the previous draft version.

Reviewer: I would be ok with saying that the client rejects the connection according to its policy. If it is content, then that is fine. Though I would strongly encourage a default posture of rejection. There are many other ways that an on path attack at this stage can be used to deny service, but leaving the protocol open to downgrade means that it is harder to defend. What if the perfectly acceptable version is broken, unbeknownst to the client?

Author: I agree with MT here, VN can already cause DoS (just send an empty VN packet) so I'd rather be more conservative unless we have a specific scenario we want to enable.

After the process of version negotiation in this document completes, the
version in use for the connection is the version that the server sent in the
`Chosen Version` field of its Version Information. That remains true even if
other versions were used in the Version field of long headers at any point in
the lifetime of the connection; endpoints MUST NOT change the version that they
consider to be in use based on the Version field of long headers as that field
could be forged by attackers.

> 

Reviewer 5: Yes! That does make the code simpler.

Author: Thanks :)

Negotiation packet, it MUST drop all subsequent Version Negotiation packets on
that connection.

Both endpoints MUST parse their peer's Version Information during the
handshake. If the Version Information was missing or if parsing it failed (for
example, if it is too short or if its length is not divisible by four), then
the endpoint MUST close the connection. If the connection was using QUIC
version 1, it MUST be closed with a transport error of type

> `TRANSPORT_PARAMETER_ERROR`.

Reviewer 4: Doesn't this mean you can't do incompatible VN when you offer VN=1?

Author: Clarified this by using a semi-colon.

If a client has reacted to a Version Negotiation packet, it MUST validate that
the server's `Other Versions` field does not contain the client's original
version, and that the client would have selected the same negotiated version if
it had received a Version Negotiation packet whose Supported Versions field had
the same contents as the server's `Other Versions` field. If any of these

> validations fail, the client MUST close the connection. If the connection was

Reviewer 4: ```suggestion
checks fail, the client MUST close the connection. If the connection was
```

using QUIC version 1, it MUST be closed with a transport error of type
`VERSION_NEGOTIATION_ERROR`. This prevents an attacker from being able to use

> forged Version Negotiation packets to force a version downgrade.

Reviewer 4: Why?

Author: Is it unclear? Can you suggest text?

Reviewer 4: This too is about the binding of "This". In this case, it's closing the connection that prevents the VN packets forcing a downgrade. I think the text I proposed above would work.

Negotiation packet, it MUST drop all subsequent Version Negotiation packets on
that connection.

Both endpoints MUST parse their peer's Version Information during the
handshake. If the Version Information was missing or if parsing it failed (for
example, if it is too short or if its length is not divisible by four), then

> the endpoint MUST close the connection; if the connection was using QUIC
> version 1, it MUST be closed with a transport error of type
> `TRANSPORT_PARAMETER_ERROR`.

Reviewer 4: ```suggestion
the endpoint MUST close the connection with a version-specific error; if the connection was using QUIC
version 1, that error MUST be `TRANSPORT_PARAMETER_ERROR`.
```
